### DISCUSSION TITLE
Embed next.js static site into another page

### DISCUSSION THREAD

**Discussion by erikmellum** [vercel/next.js#46646] [Help]
### Summary

I was given a requirement for my company to embed a next app into another HTML document. The alternative is to rewrite the app using something like Vite and CRA (in order to do client side rendering).

The approach is this:

1. `next build && next export`

2. parse the built HTML documents and extract the contents of `<head>` and `<body>` tag.

3. dynamically generate a file named `loader.js` at build time. Write this file to disk during build time. This file will contain javascript code to build a DOM node with the content of `<head>`, and another DOM node with the content of `<body>` tag. Inject these nodes into target elements on the page we are embedding into. 

4. At run time, have the source HTML document load the generated `loader.js` file. When this file is loaded, the dom nodes are created, and mounted into the target HTML document.

I am using `next build && next export` to generate the HTML for the next app. If I examine the generated .html files, I see "n" script tags in the head. When I actually _run_ the html file and interact with the app, greater than "n" scripts are present in the HEAD. For example, when one of our Table components is rendered, additional scripts are loaded into the HTML document. 

I thought this wouldn't be an issue, because I could just load the statically generated HTML using a headless browser and get the full set of script tags. However, at build time, we do not have our backend running. The script tags in the HEAD of our app are actually different when the backend is running, since it can render records in tables, etc. The way that next.js does code splitting seems to be the root cause. As you render components, javascript chunks are added into the HEAD.

Again I thought it wouldn't be an issue, because I could always just brute force this by recursively locating all scripts in the output of the static next build, and adding them to the HTML document, but that also does not work. The app loses the interactivity from react features such as useEffect, or useSwr. I'm actually not certain why this happens.

Can someone help me understand how I might detect all the scripts that are necessary during runtime, and inject them into an HTML document at build time? Is this possible? 

Is there a way to turn off code splitting? Would that solve this problem?

### Additional information

This is the post-build script, that is run after you run `next build && next export`. I start a simple http server in the out directory before running this script.

`post-build.js` 

```javascript
const fs = require('fs')
const puppeteer = require('puppeteer')
  
function createLoaderAndTestingHTML(pageObject) {
  console.log('writing test HTML page and loader')
  const loader = `
  
  function createElementFromHTML(htmlString) {
    var div = document.createElement('div')
    div.innerHTML = htmlString.trim()
  
    // Change this to div.childNodes to support multiple top-level nodes.
    return div.firstChild
  }

  const head = createElementFromHTML(\`<div>${pageObject.head}</div>\`)
  const body = createElementFromHTML(\`${pageObject.body}\`)
  
  document.querySelector('head').appendChild(head)
  document.getElementById('root-element').appendChild(body)

    
  `
  fs.writeFileSync('./out/loader.js', loader)
  fs.writeFileSync('./out/test.html', pageObject.html)
  console.log('done writing test HTML page and loader')
}

const main = async () => {
  try {
    console.log('starting')
    const browser = await puppeteer.launch({ headless: false })
    const page = await browser.newPage()
    console.log('Going to page')
    await page.goto('http://localhost:8081/supervisor/activities', { waitUntil: 'domcontentloaded' })
    const body = await page.evaluate(() => document.querySelector('body').innerHTML)
    const head = await page.evaluate(() => document.querySelector('head').innerHTML)
    const pageObject = {
      head,
      body,
      html: `<html>
      <head>
        <title>Loader test</title>
      </head>
      <body>
        <script defer src="./loader.js"></script>
        <div id="root-element"></div>
      </body>
    </html>
    `,
    }
    createLoaderAndTestingHTML(pageObject)
  } catch (e) {
    console.log('nope')
  }
}
main()
```

**Comment by jomefavourite** [comment]
Hi @erikmellum,
Were you successfully able to do this?
